### PR TITLE
Pixel Size Attributes

### LINKED ISSUE #1152: Volume pixel size
A user recently brought up the need for saving `Volume`s with a pixel size:

```
Hi,
When saving a Volume into an mrc file, the pixel size in the mrc is 0.
It is important to keep track of the correct pixel size.
For example, if we download 2660, its pixel size is 1.34A.
If we downsample it by a factor of 2, the new pixel size should be 2.68A.
This number must should in the saved mrc, as it is used, for example, by Chimera and RELION. 
```

### PR DESCRIPTION
**Pull request by garrettwrong**
Closes #1152 .

Adds pixel_size attribute to Volume and tests saving/loading default and non-default values.
Adds pixel_size attribute to Image and tests saving/loading default and non-default values.
Tests downsampling with default and non-default values.
Partially refactors Image tests to actually use fixtures (faster) and scopes them.

### CODE REVIEW

**Comment by garrettwrong** [comment]
At this point I will move onto trying to have `ImageSource` serve up `Image`s with `pixel_size` populated, and the associated tests for that.  It wasn't really part of the original issue, but might-as-well while I'm in here.

**Comment by garrettwrong** [comment]
Some more work done here.  Added `pixel_size` to synthetic volume generators, which should flow through from a user providing into simulation, or allow them to generate their own volumes with the attr.

After thinking about some options for dealing with filters, it seems like the only tricky case was `Simulation`, and its only if a user explicitly provides `pixel_size` and brings their own filters.  I've added a simple private method to check this case and raise on mismatch.  I believe all other cases should work fine with the existing filter `scale` approach; unless I've missed one.

Still need to add some more tests.

**Comment by garrettwrong** [comment]
Confirmed the pixel_size is being picked up in ChimeraX when saved from ASPIRE.

File: src/aspire/source/relion.py [hunk 1]
```diff
@@ -59,7 +59,6 @@ def __init__(
 
         self.filepath = filepath
         self.data_folder = data_folder
-        self.pixel_size = pixel_size
```
Comment (j-c-c) [inline src/aspire/source/relion.py:62]: Maybe out of scope for the PR, but should we be attempting to get the pixel size via the metadata field `_rlnImagePixelSize`?

Edit: After discussion with Garrett it looks like there was an issue about this already: #322. However, it appears to not be implemented. Maybe we should reopen to revisit after this PR.

File: src/aspire/source/micrograph.py [hunk 2]
```diff
@@ -262,8 +272,16 @@ def _images(self, indices):
...
             # Assign to array, implicitly performs casting to dtype
             micrographs[i] = micrograph.asnumpy()
+            # Assert pixel_size
+            if (
+                micrograph.pixel_size is not None
+                and micrograph.pixel_size != self.pixel_size
+            ):
+                raise NotImplementedError(
+                    f"Mismatched pixel size. {micrograph.pixel_size} defined in {self.micrograph_files[ind]}, but provided {self.pixel_size}."
+                )
```
Comment (j-c-c) [inline src/aspire/source/micrograph.py]: The wording is a little funny here. Maybe just add "angstroms" after `{micrograph.pixel_size}` and `{self.pixel_size}`.

File: src/aspire/source/micrograph.py [hunk 3]
```diff
@@ -190,11 +195,16 @@ def __init__(self, micrographs_path, dtype=None):
         # Load the first micrograph to infer shape/type
         # Size will be checked during on-the-fly loading of subsequent micrographs.
         micrograph0 = Image.load(self.micrograph_files[0])
+        if micrograph0.pixel_size is not None and micrograph0.pixel_size != pixel_size:
+            raise NotImplementedError(
+                f"Mismatched pixel size. {micrograph0.pixel_size} defined in {self.micrograph_files[0]}, but provided {pixel_size}."
+            )
```
Comment (j-c-c) [inline src/aspire/source/micrograph.py]: The wording is a little funny here. Maybe just add "angstroms" after `{micrograph.pixel_size}` and `{self.pixel_size}`.

File: src/aspire/source/simulation.py [hunk 4]
```diff
@@ -231,6 +236,29 @@ def _populate_ctf_metadata(self, filter_indices):
...
+        if self.pixel_size is None:
+            return
+
+        for f in unique_filters:
+            f_pixel_size = getattr(f, "pixel_size", None)
+            if f_pixel_size is not None and not np.isclose(
+                f_pixel_size, self.pixel_size
+            ):
+                raise ValueError(
+                    f"`Simulation.pixel_size` {self.pixel_size} does not match filter {f} {f_pixel_size}."
```
Comment (j-c-c) [inline src/aspire/source/simulation.py]: Maybe reword to `filter {f} pixel size {f_pixel_size}`

File: src/aspire/volume/volume.py [hunk 5]
```diff
@@ -242,14 +255,19 @@ def stack_reshape(self, *args):
...
             self._data.reshape(*shape, *self._data.shape[-3:]),
+            pixel_size=self.pixel_size,
             symmetry_group=self.symmetry_group,
         )
 
     def __repr__(self):
+        px_msg = "."
+        if self.pixel_size is not None:
+            px_msg = f" with pixel_size={self.pixel_size} angstroms."
+
```
Comment (j-c-c) [inline src/aspire/volume/volume.py:266]: Nice. Should we add this to the `Image` `repr` too?

File: src/aspire/volume/volume_synthesis.py [hunk 6]
```diff
@@ -39,18 +40,21 @@ class GaussianBlobsVolume(SyntheticVolumeBase):
...
-    def __init__(self, L, C, K=16, alpha=1, seed=None, dtype=np.float64):
+    def __init__(
+        self, L, C, K=16, alpha=1, pixel_size=None, seed=None, dtype=np.float64
+    ):
         """
         :param L: Resolution of the Volume(s) in pixels.
         :param C: Number of Volumes to generate.
         :param K: Number of Gaussian blobs used to construct the Volume(s).
         :param alpha: Scaling factor for variance of Gaussian blobs. Default=1.
+        :param pixel_size: Optional voxel_size in angstroms. Default=1.
```
Comment (j-c-c) [inline src/aspire/volume/volume_synthesis.py]: Default is `None`. Maybe use the same description that's in `Volume`.

File: src/aspire/volume/volume_synthesis.py [hunk 7]
```diff
@@ -168,18 +176,23 @@ class CnSymmetricVolume(GaussianBlobsVolume):
...
-    def __init__(self, L, C, order, K=16, alpha=1, seed=None, dtype=np.float64):
+    def __init__(
+        self, L, C, order, K=16, alpha=1, pixel_size=None, seed=None, dtype=np.float64
+    ):
         """
         :param L: Resolution of the Volume(s) in pixels.
         :param C: Number of Volumes to generate.
         :param order: An integer representing the cyclic order of the Volume(s).
         :param K: Number of Gaussian blobs used to construct the Volume(s).
+        :param pixel_size: Optional voxel_size in angstroms. Default=1.
```
Comment (j-c-c) [inline src/aspire/volume/volume_synthesis.py]: Default=None.

File: tests/test_image.py [hunk 8]
```diff
@@ -306,12 +318,12 @@ def testShow():
     im.show()
 
 
-def test_backproject_symmetry_group():
+def test_backproject_symmetry_group(dtype):
```
Comment (j-c-c) [inline tests/test_image.py:321]: Thanks for adding 😄

File: tests/test_volume.py [hunk 9]
```diff
@@ -291,6 +291,32 @@ def test_save_load(vols_1):
...
+        # Load saved MRC file as a Volume of dtypes single and double.
+        vols_loaded_single = Volume.load(mrcs_filepath, dtype=np.float32)
+        vols_loaded_double = Volume.load(mrcs_filepath, dtype=np.float64)
+
+        # Confirm the pixel size is loaded
+        np.testing.assert_approx_equal(
+            vols_loaded_single.pixel_size,
+            vols_2.pixel_size,
+            err_msg="Incorrect pixel size in singles.",
+        )
```
Comment (j-c-c) [inline tests/test_volume.py:322]: Maybe we should check that `vols_2.pixel_size` is correct before this comparison. This is done sort of implicitly through the synthetic volume test, but maybe doing it explicitly here for `Volume`s would be good.

**Review by j-c-c — CHANGES_REQUESTED** [review]
This looks great. Just a few things.

File: src/aspire/source/relion.py (hunk 1 above)
Comment (garrettwrong) [reply src/aspire/source/relion.py:62]: Re-opened issue.

File: src/aspire/source/micrograph.py (hunk 2 above)
Comment (garrettwrong) [reply src/aspire/source/micrograph.py]: Scientific units doesn't add anything of value wrt the actual error message, but I've added.

File: src/aspire/source/micrograph.py (hunk 3 above)
Comment (garrettwrong) [reply src/aspire/source/micrograph.py]: .

File: src/aspire/source/simulation.py (hunk 4 above)
Comment (garrettwrong) [reply src/aspire/source/simulation.py]: .

File: src/aspire/volume/volume.py (hunk 5 above)
Comment (garrettwrong) [reply src/aspire/volume/volume.py:266]: Sure.

File: src/aspire/volume/volume_synthesis.py (hunk 6 above)
Comment (garrettwrong) [reply src/aspire/volume/volume_synthesis.py]: Changed, thanks.

File: src/aspire/volume/volume_synthesis.py (hunk 7 above)
Comment (garrettwrong) [reply src/aspire/volume/volume_synthesis.py]: Changed, thanks.

File: tests/test_volume.py (hunk 9 above)
Comment (garrettwrong) [reply tests/test_volume.py:322]: Sure, added a direct test.

**Review by j-c-c — DISMISSED** [review]
Looks good!

File: src/aspire/source/micrograph.py [hunk 10]
```diff
@@ -262,8 +272,16 @@ def _images(self, indices):
                 )
             # Assign to array, implicitly performs casting to dtype
             micrographs[i] = micrograph.asnumpy()
+            # Assert pixel_size
+            if (
+                micrograph.pixel_size is not None
+                and micrograph.pixel_size != self.pixel_size
+            ):
+                raise NotImplementedError(
```
Comment (janden) [inline src/aspire/source/micrograph.py]: `ValueError`? (It's not like there's some unimplemented functionality (rescaling?) that we should have here, right?)

File: src/aspire/volume/volume.py [hunk 11]
```diff
@@ -140,7 +147,9 @@ def astype(self, dtype, copy=True):
         :return: Volume instance
         """
         return self.__class__(
-            self.asnumpy().astype(dtype, copy=copy), symmetry_group=self.symmetry_group
+            self.asnumpy().astype(dtype, copy=copy),
+            pixel_size=self.pixel_size,
```
Comment (janden) [inline src/aspire/volume/volume.py:151]: Not saying we do this now, but if (when?) we have a third field that needs to be preserved across processing steps, we may want to to think of some more compact way to do this (`…, **self._metadata_dict` where the latter is a `@property` that rounds up the things we care about into a dict and returns it?).

File: tests/test_synthetic_volume.py [hunk 12]
```diff
@@ -96,8 +100,15 @@ def test_volume_repr(vol_fixture):
...
-    _ = vol_fixture.generate()
+    """
+    Test that a volume is generated
+    and stores pixel_size when provided.
+    """
+    v = vol_fixture.generate()
+
+    # In vol_fixture, we assign pixel_size to volumes having odd voxel sizes.
+    if vol_fixture.L % 2:
+        assert v.pixel_size == 3
```
Comment (janden) [inline tests/test_synthetic_volume.py]: This will flake on some instances, no (or is 3.0 exactly representable in floating point?)?

**Review by janden — CHANGES_REQUESTED** [review]
Looks great! Thanks for tracking down the `pixel_size`s wherever they've managed to hide themselves. Just a few questions here.

File: src/aspire/source/micrograph.py (hunk 10 above)
Comment (garrettwrong) [reply src/aspire/source/micrograph.py]: Sure, changed both occurrences.

File: src/aspire/volume/volume.py (hunk 11 above)
Comment (garrettwrong) [reply src/aspire/volume/volume.py:151]: Sure, I'd be for abstracting that out in the future, maybe as it's own PR. A few classes could probably use some cleanup like that at this point.

File: tests/test_synthetic_volume.py (hunk 12 above)
Comment (garrettwrong) [reply tests/test_synthetic_volume.py]: I think 3 is representable in floats and doubles, but you are right that it is better to compare via closeness just in case.  Changed to `assert_approx_equal`.

**Comment by garrettwrong** [comment]
Rebased to resolve minor downsample merge ocnflicts.

File: src/aspire/source/micrograph.py (hunk 10 above)
Comment (janden) [reply src/aspire/source/micrograph.py]: Thanks!

File: tests/test_synthetic_volume.py (hunk 12 above)
Comment (janden) [reply tests/test_synthetic_volume.py]: Great. Thanks!

**Review by janden — APPROVED** [review]
